### ISSUE TITLE
Never returning branch is expected to return something by compiler, instead of never

### ISSUE THREAD

**Issue by 3n-mb** [Microsoft/TypeScript#13957]
**TypeScript Version:**  2.1.4
`"strictNullChecks": true`

**Code**

```ts

function justThrow(): never {
	throw new Error('We throw');
}

function makeString(n: number): string {
	if (n > 0) {
		return `${n}`;
	} else {
		justThrow();
	}
}

```

**Expected behavior:**
The *never returning* branch should not be required to have return, and code should compile, by this logic.

**Actual behavior:**
Error is reported:
`
error TS2366: Function lacks ending return statement and return type does not include 'undefined'.
`
And the same thing happens when *never returning* branch is set in a try-catch structure.

**Comment by RyanCavanaugh** [comment]
Same comment in #13958 applies here
